### DISCUSSION TITLE
implementing JTAG protocol using PIO state machine

### DISCUSSION THREAD

**Discussion by Prakhar005** [micropython/micropython#9371] [Using MicroPython] [answered]
Hello,

I am trying to implement JTAG protocol using a PIO state machine. I need to generate two signals- TDI and TCK and have assigned the pins for them in my state machine. During one of the states, I need to serially send out data on TDI pin, along with the clock on a TCK pin. I am able to create a clock by using the set() command for toggling the TCK pin. However, I am unable to figure out how to serially shift out a bit on every TCK cycle on the TDI pin. 
For now I am working with a 32-bit data stored in OSR that needs to be shifted out every TCK cycle. Appreciate any help!

**Comment by robert-hh** [comment, marked as answer]
You should shift out the data with out(pin, 1) and control the CLK pin with sideset on the set command. The 1 in the out command outputs just une bit and shift the reguister, such that you can loop through the word. See e.g. https://github.com/robert-hh/SSD1963-TFT-Library-for-PyBoard-and-RP2040/blob/master/rp2040/tft_pio.py, function pio_data_write, where three bus control lines are set with the data.

**Reply by Prakhar005** [reply]
Thank you very much! I was unfamiliar with the sideset command. I implemented the method you suggested and it worked perfectly. I had a follow up question regarding the same. I implement the code for shifting data with every CLK cycle as follows:

@asm_pio(out_init = PIO.OUT_LOW, sideset_init=PIO.OUT_LOW)
def jtag_write():
    out(pins,1)   .side(0).      % shift 1 bit out while keeping CLK low
    nop()            .side(1)        % set CLK to high


Does the second instruction where I use a nop() and sideset command together, count as an instruction in the instruction memory. The reason I ask is because we can have maximum 32 instructions in a PIO and I want to ensure that I am not wasting instructions.
Thank you for the help!

**Comment by robert-hh** [comment]
The nop() is an instruction. If you decode the binary of nop(), you'll see that it is in fact a mov of a register to itself.

**Reply by Prakhar005** [reply]
Thank you Robert. That makes a lot of sense. Seems like I will have to be very cautious in using instructions, since JTAG state machine is quite large. Thank you for the help!
